File: creational/dart/builder_structured/director.dart
import 'interfaces.dart';

class ConstructionDirector {
  HouseBuilder? _builder;
  
  void setBuilder(HouseBuilder builder) {
    _builder = builder;
  }
  
  void buildBasicHouse() {
    _builder!.buildFoundation();
    _builder!.buildStructure();
    _builder!.buildRoof();
    _builder!.buildInterior();
  }
  
  void buildFullHouse() {
    _builder!.buildFoundation();
    _builder!.buildStructure();
    _builder!.buildRoof();
    _builder!.buildInterior();
    _builder!.buildGarage();
    _builder!.buildGarden();
    _builder!.buildSwimmingPool();
  }
  
  void buildCustomHouse(bool withGarage, bool withGarden, bool withPool) {
    _builder!.buildFoundation();
    _builder!.buildStructure();
    _builder!.buildRoof();
    _builder!.buildInterior();
    if (withGarage) _builder!.buildGarage();
    if (withGarden) _builder!.buildGarden();
    if (withPool) _builder!.buildSwimmingPool();
  }
}

File: creational/typescript/builder_structured/director.ts
import { HouseBuilder } from './interfaces';

export class ConstructionDirector {
  private builder: HouseBuilder | null = null;
  
  setBuilder(builder: HouseBuilder): void {
    this.builder = builder;
  }
  
  buildBasicHouse(): void {
    if (!this.builder) return;
    this.builder.buildFoundation();
    this.builder.buildStructure();
    this.builder.buildRoof();
    this.builder.buildInterior();
  }
  
  buildFullHouse(): void {
    if (!this.builder) return;
    this.builder.buildFoundation();
    this.builder.buildStructure();
    this.builder.buildRoof();
    this.builder.buildInterior();
    this.builder.buildGarage();
    this.builder.buildGarden();
    this.builder.buildSwimmingPool();
  }
  
  buildCustomHouse(withGarage: boolean, withGarden: boolean, withPool: boolean): void {
    if (!this.builder) return;
    this.builder.buildFoundation();
    this.builder.buildStructure();
    this.builder.buildRoof();
    this.builder.buildInterior();
    if (withGarage) this.builder.buildGarage();
    if (withGarden) this.builder.buildGarden();
    if (withPool) this.builder.buildSwimmingPool();
  }
}
